src/python/pants/backend/python/goals/lockfile.py

```diff
@@ -268,10 +270,50 @@ async def setup_user_lockfile_requests(
     )
 
 
+@dataclass(frozen=True)
+class PythonSyntheticLockfileTargetsRequest(SyntheticTargetsRequest):
+    """Register the type used to create synthetic targets for Python lockfiles.
+
+    As the paths for all lockfiles are known up-front, we set the `path` field to
+    `SyntheticTargetsRequest.SINGLE_REQUEST_FOR_ALL_TARGETS` so that we get a single request for all
+    our synthetic targets rather than one request per directory.
+    """
+
+    path: str = SyntheticTargetsRequest.SINGLE_REQUEST_FOR_ALL_TARGETS
+
+
+@rule
+async def python_lockfile_synthetic_targets(
+    request: PythonSyntheticLockfileTargetsRequest,
+    python_setup: PythonSetup,
+) -> SyntheticAddressMaps:
+    if not python_setup.enable_resolves:
+        return SyntheticAddressMaps()
+
+    resolves = [
+        (os.path.dirname(lockfile), os.path.basename(lockfile), name)
+        for name, lockfile in python_setup.resolves.items()
+    ]
+    return SyntheticAddressMaps.for_targets_request(
+        request,
+        [
+            (
+                os.path.join(spec_path, "BUILD.python-lockfiles"),
+                tuple(
+                    TargetAdaptor("_lockfiles", name=name, sources=[lockfile])
+                    for _, lockfile, name in lockfiles
```

Comment:
To fix #18404 I think we'll need to check here that the lockfile exists?

Also, I'm not sure we want to use the resolve name as the target name, that may collide with another target name in the dir. We should probably prefix it some something.

Reply:
> To fix #18404 I think we'll need to check here that the lockfile exists?

I think it would be preferable if we could change `_lockfiles.source` to not require it's source glob to match regardless of the setting for the global `unmatched_build_file_globs` option.

> Also, I'm not sure we want to use the resolve name as the target name, that may collide with another target name in the dir. We should probably prefix it some something.

That have already been addressed by #17365

Reply:
Partly, but at the const of extra complication, and burden on the user, when there is a collision. We could make the problem go away entirely by synthesizing a less obvious name. Even just a `_` prefix would avoid the issue ~entirely.

Reply:
"Resolve by either renaming the shadowing target, the resolve 'python-default' or moving the target or the lockfile to another directory."

The user just named their "real" target something reasonable, like "pytest" (or left the name off entirely so it defaulted to the directory name, say "pytest") and now we've decided to squat on that name, and made work for them to fix the situation. A synthetic target should have a name that is unlikely to collide, because we should defer to the user, not the other way around.

Reply:
Ah, ok. yeah, a `_` prefix could be reasonable and close enough to my desired target name to be sensible.